precompiles/parachain-staking/src/lib.rs

```diff
@@ -57,469 +62,635 @@ where
 		target_gas: Option<u64>,
 		context: &Context,
 	) -> Result<PrecompileOutput, ExitError> {
-		log::trace!(target: "staking-precompile", "In parachain staking wrapper");
+		let input = InputReader::new(input)?;
+
+		match input.selector() {
+			// constants
+			[0xc9, 0xf5, 0x93, 0xb2] => Self::min_nomination(target_gas),
+			// storage getters
+			[0x97, 0x99, 0xb4, 0xe7] => Self::points(input, target_gas),
+			[0x4b, 0x1c, 0x4c, 0x29] => Self::candidate_count(target_gas),
+			[0x0a, 0xd6, 0xa7, 0xbe] => Self::collator_nomination_count(input, target_gas),
+			[0xda, 0xe5, 0x65, 0x9b] => Self::nominator_nomination_count(input, target_gas),
+			// role verifiers
+			[0x8e, 0x50, 0x80, 0xe7] => Self::is_nominator(input, target_gas),
+			[0x85, 0x45, 0xc8, 0x33] => Self::is_candidate(input, target_gas),
+			[0x8f, 0x6d, 0x27, 0xc7] => Self::is_selected_candidate(input, target_gas),
+			// runtime methods (setters)
+			[0x0a, 0x1b, 0xff, 0x60] => Self::join_candidates(input, target_gas, context),
+			[0x72, 0xb0, 0x2a, 0x31] => Self::leave_candidates(input, target_gas, context),
+			[0x76, 0x7e, 0x04, 0x50] => Self::go_offline(target_gas, context),
+			[0xd2, 0xf7, 0x3c, 0xeb] => Self::go_online(target_gas, context),
+			[0x28, 0x9b, 0x6b, 0xa7] => Self::candidate_bond_less(input, target_gas, context),
+			[0xc5, 0x7b, 0xd3, 0xa8] => Self::candidate_bond_more(input, target_gas, context),
+			[0x49, 0xdf, 0x6e, 0xb3] => Self::nominate(input, target_gas, context),
+			[0xb7, 0x1d, 0x21, 0x53] => Self::leave_nominators(input, target_gas, context),
+			[0x4b, 0x65, 0xc3, 0x4b] => Self::revoke_nomination(input, target_gas, context),
+			[0xf6, 0xa5, 0x25, 0x69] => Self::nominator_bond_less(input, target_gas, context),
+			[0x97, 0x1d, 0x44, 0xc8] => Self::nominator_bond_more(input, target_gas, context),
+			_ => Err(error(
+				"No parachain-staking wrapper method at given selector",
+			)),
+		}
+	}
+}
 
-		// Basic sanity checking for length
-		// https://solidity-by-example.org/primitives/
+impl<Runtime> ParachainStakingWrapper<Runtime>
+where
+	Runtime: parachain_staking::Config + pallet_evm::Config + frame_system::Config,
+	Runtime::AccountId: From<H160>,
```

Comment:
Probably for another PR, but I just wanted to share this idea. I believe we can get rid of the `From<H160>` bound by using https://github.com/paritytech/frontier/blob/master/frame/evm/src/lib.rs#L195-L197 instead.

Reply:
The `InputReader::read_address` function is generic over `AccountId: From<H160>` so I think it is useful for that reason.